xtrack/monitors/bunch_monitor.h

```diff
@@ -0,0 +1,71 @@
+// ##################################
+// Bunch Monitor
+// 
+// Author: Philipp Niedermayer
+// Date: 2023-08-14
+// Edit: 2024-06-12
+// ##################################
+
+
+#ifndef XTRACK_BEAM_SIZE_MONITOR_H
+#define XTRACK_BEAM_SIZE_MONITOR_H
+
+#if !defined( C_LIGHT )
+    #define   C_LIGHT ( 299792458.0 )
+#endif /* !defined( C_LIGHT ) */
+
+/*gpufun*/
+void BeamSizeMonitor_track_local_particle(BeamSizeMonitorData el, LocalParticle* part0){
+
+    // get parameters
+    int64_t const start_at_turn = BeamSizeMonitorData_get_start_at_turn(el);
+    int64_t particle_id_start = BeamSizeMonitorData_get_particle_id_start(el);
+    int64_t particle_id_stop = particle_id_start + BeamSizeMonitorData_get_num_particles(el);
+    // Should we do explicit casting to float (?) 
+    int64_t const harmonic = BeamSizeMonitorData_get_harmonic(el);
+    double const frev = BeamSizeMonitorData_get_frev(el);
+    
+    BeamSizeMonitorRecord record = BeamSizeMonitorData_getp_data(el);
+
+    int64_t max_slot = BeamSizeMonitorRecord_len_count(record);
+
+    //start_per_particle_block(part0->part)
+
+        int64_t particle_id = LocalParticle_get_particle_id(part);
+        if (particle_id_stop < 0 || (particle_id_start <= particle_id && particle_id < particle_id_stop)){
+
+            // zeta is the absolute path length deviation from the reference particle: zeta = (s - beta0*c*t)
+            // but without limits, i.e. it can exceed the circumference (for coasting beams)
+            // as the particle falls behind or overtakes the reference particle
+            double const zeta = LocalParticle_get_zeta(part);
+            double const at_turn = LocalParticle_get_at_turn(part);
+            double const beta0 = LocalParticle_get_beta0(part);
+
+            // compute sample index
+            int64_t slot = round( harmonic * ( (at_turn-start_at_turn) - frev * zeta/beta0/C_LIGHT ));
```

Comment:
I'm thinking maybe you could use the circumference `line.get_length()` instead to simplify the slot calculation.
In principle, the zeta correction is not required if the monitor is only used for bunched beams and h≥1. But I think it's good to keep to properly handle RF gymnastics like de- and re-bunching.

Reply:
I agree, maybe it could be useful in the future.